### ISSUE TITLE
Process.WaitForExit() hangs forever even though the process has exited in some cases

### ISSUE THREAD

**Issue by madelson** [dotnet/corefx#36783]
**Setup**

Launch a process with stream redirection where that process itself launches a long-running process:

```
// in the host process (process1.exe)
var process2 = new Process
{
    StartInfo = 
    {
        FileName = "process2.exe"
        UseShellExecute = false,
        RedirectStandardOutput = true
    }
};
process2.OutputDataReceived += (o, e) => Console.WriteLine(e.Data);
process2.Start();
process2.BeginOutputReadLine();
process2.WaitForExit(); // never returns, even though process2 does exit fairly quickly
return; // returns from Main(), exiting

// in process2.exe
Console.WriteLine("Launching process3.exe!");
var process3 = new Process { StartInfo = { FileName = "process3.exe", UseShellExecute = false } };
process3.Start();
process3.Dispose();

// in process3.exe
while (true) // loop basically forever performing background tasks
{
     ...
}
```

**Expected Behavior**
Process1 receives the output of Process2. Once Process2 has exited, Process1 can move on from its `WaitForExit()` call. Meanwhile Process3 continues to run.

**Actual Behavior**
`WaitForExit()` never returns.

**Notes**
I believe the problem is that `WaitForExit()` waits for EOF to be returned from redirected streams that are being consumed asynchronously. However, I think that Process3 is somehow inheriting the stdio handles from Process2 which prevents EOF from being sent even after Process2 exits.

Note that unlike #29699 and #31713 , I am observing the behavior on Windows. I'm not sure if the same would happen on Linux.

**Comment by tmds** [comment]
> I believe the problem is that WaitForExit() waits for EOF to be returned from redirected streams that are being consumed asynchronously. However, I think that Process3 is somehow inheriting the stdio handles from Process2 which prevents EOF from being sent even after Process2 exits.

Yes, Process2 and Process3 share the same standard out. What happens when you `RedirectStandardOutput = true` for Process3?

**Comment by tmds** [comment]
WaitForExit involves waiting for the output to reach EOF, which won't happen because Process3 is keeping it alive.

https://github.com/dotnet/corefx/blob/007571b90c3cc555cd30f3b467b87c22f9f665e0/src/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs#L210-L228

**Comment by danmosemsft** [comment]
@tmds the code above is Unix, the report is Windows (you probably intended this but just pointing out in case)

Looks like the code has a typo "WaitUtilEOF"

**Comment by tmds** [comment]
For Windows, we do the WaitUtilEOF also:

https://github.com/dotnet/corefx/blob/007571b90c3cc555cd30f3b467b87c22f9f665e0/src/System.Diagnostics.Process/src/System/Diagnostics/Process.Windows.cs#L173-L183

**Comment by tmds** [comment]
@danmosemsft what is the rationale for the WaitUtilEOF? From the code I think it is not really needed on Unix, but it seems to serve a role on Windows: the handle is Disposed _after_ the WaitUtilEOF.

@madelson these two options should make it do what you want:
- `RedirectStandardOutput&Error = true`  of Process3
- `Close()` `StandardOutput&Error` of Process2

**Comment by danmosemsft** [comment]
@tmds I do not know, and an not able to examine the code at the moment. Perhaps @krwq  @stephentoub know.

**Comment by madelson** [comment]
@tmds I agree that this would probably fix. The problem is that in this case I don't control the source of process #2; that's just what it does.

**Comment by tmds** [comment]
@madelson you can do the Close, that should work too.

**Comment by danmosemsft** [comment]
@madelson can you please confirm this fixes it?

**Comment by madelson** [comment]
@danmosemsft I will confirm soon when I next get a chance to retry the code. Thinking about this more I’m not convinced it will fix

**Comment by tmds** [comment]
> Perhaps @krwq @stephentoub know.

@krwq @stephentoub do you know why WaitForExit reads standard out and error until EOF?

**Comment by krwq** [comment]
@tmds got no context on that, perhaps some simplification to avoid races between events and exit. If the question is just about linux then I suspect the code was ported from Windows and it's there for parity - if you think it shouldn't be there I think it makes sense to try to remove it and try running tests for some time in a loop and perhaps add some more targeted to see if there are any issues.

**Comment by madelson** [comment]
@krwq I imagine that partly what it's trying to achieve is to give you some reasonable way of knowing when everything is done and no more data is coming. With WaitForExit() not returning until this happens, it allows you to set up data received events, wait for the process to finish, and then move on.

**Comment by tmds** [comment]
@madelson did the Close of process2 stdout/stderr work for you?

>  I think it makes sense to try to remove it and try running tests for some time in a loop and perhaps add some more targeted to see if there are any issues.

We should keep it in both, or remove it in both, to have the same behavior. I wonder if on Windows the Dispose of the handle would affect the behavior of standard out and standard error if they were not yet read until EOF. I'll make a PR that removes the readuntileof and see if some tests fail.

**Comment by tmds** [comment]
Without the WaitUntilEOF tests fail that depend on OutputdataReceived being called before WaitForExit returns, for example:

https://github.com/dotnet/corefx/blob/4e1b64e1ab06202e9aa451a0b320b7fd4ed11519/src/System.Net.Ping/tests/FunctionalTests/UnixPingUtilityTests.cs#L74-L75

https://github.com/dotnet/corefx/blob/4e1b64e1ab06202e9aa451a0b320b7fd4ed11519/src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs#L88

https://github.com/dotnet/corefx/blob/4e1b64e1ab06202e9aa451a0b320b7fd4ed11519/src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs#L417

**Comment by stephentoub** [comment]
Yeah, the "eof" part is a bit of a misnomer. It's really waiting until all background processing completes.

**Comment by madelson** [comment]
@stephentoub perhaps a solution would be for the process streams to support cancellation / timeout. If they did, then something could be set up where the process exiting cancels any outstanding reads (although even in that case you might have to try one more read on the stream to be sure that there's nothing left in the pipe).

**Comment by Peperud** [comment]
Is there a workaround for this?

**Comment by danmosemsft** [comment]
@Peperud  see above https://github.com/dotnet/corefx/issues/36783#issuecomment-482780224

**Comment by madelson** [comment]
> Close() StandardOutput&Error of Process2

> did the Close of process2 stdout/stderr work for you?

@tmds I don't think this is a workaround because I am trying to capture the standard output of process2. If I close stdout before waiting for exit, I won't get the output. If I wait for exit before closing, then the program hangs.

**Comment by tmds** [comment]
> @tmds I don't think this is a workaround because I am trying to capture the standard output of process2. If I close stdout before waiting for exit, I won't get the output. If I wait for exit before closing, then the program hangs.

Then you need to [redirect the output of process 3](https://github.com/dotnet/corefx/issues/36783#issuecomment-482780224), otherwise you can't know when process2 output ends.

**Comment by madelson** [comment]
> Then you need to redirect the output of process 3, otherwise you can't know when process2 output ends.

@tmds is this because processes 2 and 3 share a console? Otherwise, I don't see how one would affect the other.

**Comment by tmds** [comment]
Yes, when starting process 2, process1 creates a pipe (RedirectStandardOutput) and both process2 and process3 stdout write to it.

**Comment by Peperud** [comment]
Am I reading this right? I must redirect the output of any process launched down the chain? 

I can only ensure compliance for the processes that I start directly. But if any of them launches a process of their own I'll be doomed? 

Doesn't sound quite right to me.

**Comment by tmds** [comment]
> Am I reading this right? I must redirect the output of any process launched down the chain?

You are reading it right.
If you want to distinguish between the output of a direct child, and their children, those outputs need to write to different pipes.

> I can only ensure compliance for the processes that I start directly. But if any of them launches a process of their own I'll be doomed?

It's a bug when a program launches a child that keeps stdout open when the child isn't going to provide output (especially if that child out-lives the parent process).

Related API proposal: https://github.com/dotnet/corefx/issues/35685

**Comment by Peperud** [comment]
> You are reading it right.
> If you want to distinguish between the output of a direct child, and their children, those outputs need to write to different pipes.

Understood, but I'm not concerned with distinguishing the outputs here. 

> 
> It's a bug when a program launches a child that keeps stdout open when the child isn't going to provide output (especially if that child out-lives the parent process).
> 
Not sure if this applies to what I see. No child down the line outlives its parent. All children (potentially) write output. Yet, the parent hangs on ```WaitForExit```

> Related API proposal: #35685

**Comment by tmds** [comment]
>  I'm not concerned with distinguishing the outputs here

I mean you don't want grand children keeping standard output open (even if they don't actively write to it).

> No child down the line outlives its parent. All children (potentially) write output. Yet, the parent hangs on WaitForExit

If the child and everything down the line has exited, WaitForExit should return.
Are you running in a container?
Can you reproduce this?

**Comment by Peperud** [comment]
> If the child and everything down the line has exited, WaitForExit should return.
> Are you running in a container?

Yes. 
> Can you reproduce this?

I've moved away to a different arrangement, but if I get some time I'll try to reproduce again.

**Comment by tmds** [comment]
> Are you running in a container?
Yes

WaitForExit can be quirky in containers when there is no init process. The init process reaps orphaned child processes. You can add one with `docker run` by adding an `--init` argument.
